In [101]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression

# Import Data Bursa Transfer 2000-2018

Dataset yang saya gunakan adalah data bursa transfer sepak bola dari berbagai negara untuk tahun 2000-2018. Melalui dataset ini, saya akan coba menganalisis dan mengklasifikasikan pemain mana yang berpotensi overpriced dengan feature yang cukup common

In [102]:
dataset = pd.read_csv('C:/Users/Bangga Rakana Adian/Desktop/Kuliah/python/TopTransfer1.csv')

In [103]:
dataset

Name            Position  Age        Team_from  \
0                   Luís Figo        Right Winger   27     FC Barcelona   
1               Hernán Crespo      Centre-Forward   25            Parma   
2               Marc Overmars         Left Winger   27          Arsenal   
3           Gabriel Batistuta      Centre-Forward   31       Fiorentina   
4              Nicolas Anelka      Centre-Forward   21      Real Madrid   
5               Rio Ferdinand         Centre-Back   22         West Ham   
6            Flávio Conceicao    Central Midfield   26   Dep. La Coruña   
7              Savo Milosevic      Centre-Forward   26    Real Zaragoza   
8             David Trézéguet      Centre-Forward   22           Monaco   
9               Claudio López      Centre-Forward   25      Valencia CF   
10    Jimmy Floyd Hasselbaink      Centre-Forward   28  Atlético Madrid   
11               Gerard López    Central Midfield   21      Valencia CF   
12                      Lucas      Centre-Forward   21      Atlético-PR   
13                Pablo Aimar  Attacking Midfield   21      River Plate   
14              Wálter Samuel         Centre-Back   22     Boca Juniors   
15              Shabani Nonda      Centre-Forward   23    Stade Rennais   
16               Robbie Keane      Centre-Forward   19    Coventry City   
17                  José Mari      Centre-Forward   21  Atlético Madrid   
18            Jonathan Zebina         Centre-Back   21  Cagliari Calcio   
19                    Émerson  Defensive Midfield   24  Bay. Leverkusen   
20             Tore André Flo      Centre-Forward   27          Chelsea   
21              Sergiy Rebrov      Second Striker   26      Dynamo Kyiv   
22             Angelo Peruzzi          Goalkeeper   30            Inter   
23              Diego Tristán      Centre-Forward   24     RCD Mallorca   
24            Sylvain Wiltord      Second Striker   26      G. Bordeaux   
25           Fernando Redondo  Defensive Midfield   31      Real Madrid   
26                     Jardel      Centre-Forward   26         FC Porto   
27                 Nuno Gomes      Centre-Forward   24          Benfica   
28           Sérgio Conceição        Right Winger   25            Lazio   
29                    Alfonso      Centre-Forward   27       Real Betis   
...                       ...                 ...  ...              ...   
4670         Benedikt Höwedes         Centre-Back   30    FC Schalke 04   
4671           Gonzalo Castro    Central Midfield   31    Bor. Dortmund   
4672              Sergi Gómez         Centre-Back   26    Celta de Vigo   
4673              Adam Masina           Left-Back   24          Bologna   
4674       Christophe Hérelle         Centre-Back   25           Troyes   
4675            Firmin Mubele      Centre-Forward   24    Stade Rennais   
4676      Florent Hadergjonaj          Right-Back   23    FC Ingolstadt   
4677       Nemanja Maksimovic    Central Midfield   23      Valencia CF   
4678              Juan Iturbe        Right Winger   25          AS Roma   
4679                  Sidcley           Left-Back   25      Atlético-PR   
4680                 Fabrício      Centre-Forward   28     Portimonense   
4681             Johan Mojica           Left-Back   25   Rayo Vallecano   
4682                   Petros    Central Midfield   29        São Paulo   
4683         Christian Kouamé      Centre-Forward   20       Cittadella   
4684         Luca Waldschmidt      Centre-Forward   22     Hamburger SV   
4685             Marc Muniesa         Centre-Back   26       Stoke City   
4686          Pietro Iemmello      Centre-Forward   26         Sassuolo   
4687             Felipe Vizeu      Centre-Forward   21         Flamengo   
4688         Marco Tumminello      Centre-Forward   19          AS Roma   
4689             Jens Odgaard      Centre-Forward   19            Inter   
4690            Marco Carraro  Defensive Midfield   20            Inter   
4691           Jonathan Ikoné  Attacking Midfield 

# Exploratory Data Analysis

In [104]:
dataset['League_from'].value_counts().head(10)

Premier League    608
Serie A           602
Ligue 1           428
LaLiga            418
1.Bundesliga      265
Série A           199
Championship      197
Eredivisie        190
Liga NOS          178
Premier Liga      169
Name: League_from, dtype: int64

In [105]:
dataset['League_to'].value_counts().head(10)

Premier League    1256
Serie A            739
LaLiga             525
1.Bundesliga       422
Ligue 1            397
Premier Liga       328
Süper Lig          143
Liga NOS           127
Super League       122
Eredivisie         108
Name: League_to, dtype: int64

Premier League Inggris menjadi liga yang paling aktif di bursa transfer

Slicing data untuk melihat masing-masing aktivitas transfer domestik dan transfer lintas liga

In [106]:
domestik = dataset[dataset['League_from'] == dataset['League_to']]
len(domestik)

1530

In [107]:
lintas = dataset[dataset['League_from'] != dataset['League_to']]
len(lintas)

3170

In [108]:
domestik['League_from'].value_counts().head()

Premier League    374
Serie A           359
Ligue 1           159
LaLiga            147
1.Bundesliga      136
Name: League_from, dtype: int64

In [109]:
domestik['Team_from'].value_counts().head()

Spurs        38
Inter        36
Genoa        33
Parma        32
Liverpool    32
Name: Team_from, dtype: int64

In [110]:
domestik['Team_to'].value_counts().head()

Juventus    53
Inter       45
AC Milan    31
AS Roma     31
Man City    29
Name: Team_to, dtype: int64

In [111]:
lintas['League_from'].value_counts().head()

LaLiga            271
Ligue 1           269
Serie A           243
Premier League    234
Série A           190
Name: League_from, dtype: int64

In [112]:
lintas['Team_from'].value_counts().head()

FC Porto           56
Benfica            52
Atlético Madrid    45
Real Madrid        41
River Plate        39
Name: Team_from, dtype: int64

In [113]:
lintas['League_to'].value_counts().head()

Premier League    882
Serie A           380
LaLiga            378
1.Bundesliga      286
Premier Liga      252
Name: League_to, dtype: int64

In [114]:
lintas['Team_to'].value_counts().head()

Chelsea      75
Spurs        68
Man City     65
Liverpool    60
Arsenal      54
Name: Team_to, dtype: int64

Aktivitas transfer lintas liga lebih tinggi dibandingkan dengan aktivitas transfer liga domestik secara global.
Tim-tim Serie A Italia mendominasi aktivitas transfer liga domestik. Dua tim dari Liga NOS Portugal yaitu FC Porto dan Benfica menjadi yang paling sering memprovide pemainnya ke liga lain. Tim-tim papan atas Premier League Inggris menjadi yang paling sering mendatangkan pemain asing, jauh mengungguli tim dari liga lain

In [115]:
print('Market val rata-rata domestik = {}\nMarket val rata-rata lintas = \t{}'.format((domestik['Market_value'].mean()), ((lintas['Market_value'].mean())))) 

Market val rata-rata domestik = 8061960.784313725
Market val rata-rata lintas = 	7912924.29022082


In [116]:
print('Transfer Fee rata-rata domestik = {}\nTransfer Fee rata-rata lintas =   {}'.format((domestik['Transfer_fee'].mean()), ((lintas['Transfer_fee'].mean())))) 

Transfer Fee rata-rata domestik = 9549379.08496732
Transfer Fee rata-rata lintas =   9398455.835962145


In [117]:
print('Market Val total domestik = {}\nMarket Val total lintas =   {}'.format((domestik['Market_value'].sum()), (lintas['Market_value'].sum())))

Market Val total domestik = 12334800000
Market Val total lintas =   25083970000


In [118]:
print('Transfer Fee total domestik = {}\nTransfer Fee total lintas =   {}'.format((domestik['Transfer_fee'].sum()), (lintas['Transfer_fee'].sum())))

Transfer Fee total domestik = 14610550000
Transfer Fee total lintas =   29793105000


In [119]:
domestik['Season'].value_counts().head()

2011-2012    96
2009-2010    92
2013-2014    92
2008-2009    89
2001-2002    88
Name: Season, dtype: int64

In [120]:
lintas['Season'].value_counts().head()

2004-2005    179
2000-2001    178
2014-2015    176
2003-2004    172
2005-2006    171
Name: Season, dtype: int64

Secara keseluruhan, transfer lintas liga jauh mengungguli transfer domestik untuk nilai pasar dan nilai beli. Namun, rata-ratanya kalah unggul dibandingkan dengan transfer domestik. Dapat dipastikan bahwa nilai yang dihargai untuk masing-masing pemainnya jauh lebih kecil dibandingkan dengan transfer domestik. 

Destinasi awal dan akhir liga berpengaruh terhadap selisih nilai pasar dan nilai beli pemain sepak bola

In [121]:
positioning = dataset['Position']
post = positioning.replace(('Centre-Forward', 'Second Striker', 'Right Winger', 'Left Winger', 'Forward', 'Right Midfield', 'Attacking Midfield', 'Left Midfield', 'Central Midfield', 'Defensive Midfield', 'Midfielder', 'Right-Back', 'Centre-Back', 'Left-Back', 'Sweeper', 'Defender', 'Goalkeeper'), ('FW','FW','FW','FW','FW','MD','MD','MD','MD','MD','MD','DF','DF','DF','DF','DF','GK'))
dataset['Role'] = post

In [122]:
param = ['Age', 'Role', 'Season']
a = dataset[dataset['Transfer_fee'] > dataset['Market_value']][param]
b = a['Role'].value_counts()
c = dataset['Role'].value_counts()
d = a['Age'].value_counts().head()
s = a['Season'].value_counts().head()
print('Footballers average age who were overprice = \n{}\n\nAmounts of Role in dataset = \n{}\n\nAmounts of Role who were overpriced = \n{}\n\nSeasons = \n{}'.format(d,c,b,s))

Footballers average age who were overprice = 
24    270
23    265
25    255
22    243
21    188
Name: Age, dtype: int64

Amounts of Role in dataset = 
FW    1923
MD    1475
DF    1122
GK     180
Name: Role, dtype: int64

Amounts of Role who were overpriced = 
FW    950
MD    626
DF    525
GK     76
Name: Role, dtype: int64

Seasons = 
2016-2017    208
2017-2018    196
2015-2016    191
2014-2015    169
2009-2010    169
Name: Season, dtype: int64


Rata-rata usia pemain sepak bola yang overpriced adalah ~24 tahun. Di dalam dataset, dapat terlihat aktivitas transfer untuk masing-masing posisi, dimana pemain penyerang paling banyak dilibatkan dalam kegiatan transfer. Jumlah pemain overpriced di posisi penyerang jauh mengungguli pemain posisi lain. Dalam 3 tahun terakhir, jumlah pemain overpriced terus meningkat

In [123]:
fw = dataset[dataset['Role'] == 'FW']['Transfer_fee'].mean()
md = dataset[dataset['Role'] == 'MD']['Transfer_fee'].mean()
df = dataset[dataset['Role'] == 'DF']['Transfer_fee'].mean()
gk = dataset[dataset['Role'] == 'GK']['Transfer_fee'].mean()
print('Average transfer fee of attackers = \n{}\n\nAverage transfer fee of midfielders = \n{}\n\nAverage transfer fee of defenders = \n{}\n\nAverage transfer fee of goalkeepers = \n{}\n\n'.format(fw,md,df,gk))

Average transfer fee of attackers = 
10355990.639625585

Average transfer fee of midfielders = 
9389606.779661017

Average transfer fee of defenders = 
8259656.862745098

Average transfer fee of goalkeepers = 
7622666.666666667




Posisi dan usia mempengaruhi adanya selisih nilai pasar dan nilai beli pemain sepak bola

# Feature Engineering

In [124]:
vd = dataset['Transfer_fee'] - dataset['Market_value']
vd[vd > 0] = 1
vd[vd <= 0] = 0
dataset['Category'] = vd

In [125]:
mig = dataset['League_from'] == dataset['League_to']
migration = mig.replace((True, False), (1,0))
dataset['Migration'] = migration

In [126]:
age_max = dataset['Age'].max()
age_min = dataset['Age'].min()
age = (dataset['Age'] - age_min) / (age_max - age_min)


In [127]:
musim = dataset['Season']
season = musim.replace(('2000-2001', '2001-2002', '2002-2003', '2003-2004', '2004-2005', '2005-2006', '2006-2007', '2007-2008', '2008-2009', '2009-2010', '2010-2011', '2011-2012', '2012-2013', '2013-2014', '2014-2015', '2015-2016', '2016-2017', '2017-2018', '2018-2019'), (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18, 19))
season_max = season.max()
season_min = season.min()
sea = (season - season_min) / (season_max - season_min)

In [128]:
roles = dataset['Role']
role = roles.replace(('FW', 'MD', 'DF', 'GK'), (4,3,2,1))
role_max = role.max()
role_min = role.min()
Role = (role - role_min) / (role_max - role_min)

In [129]:
Role = pd.Series(Role)
Age = pd.Series(age)
Season = pd.Series(sea)
Value_difference = pd.Series(vd)
Migration = pd.Series(migration)
dt = pd.DataFrame({'Role': Role, 'Age': Age,'Season': Season,'Category': Value_difference, 'Migration': Migration})
dataset2 = dt[['Age', 'Season','Role','Migration','Category']]
dataset2

Age  Season      Role  Migration  Category
0     0.771429     0.0  1.000000          1         0
1     0.714286     0.0  1.000000          1         0
2     0.771429     0.0  1.000000          0         0
3     0.885714     0.0  1.000000          1         0
4     0.600000     0.0  1.000000          0         0
5     0.628571     0.0  0.333333          1         0
6     0.742857     0.0  0.666667          1         0
7     0.742857     0.0  1.000000          0         0
8     0.628571     0.0  1.000000          0         0
9     0.714286     0.0  1.000000          0         0
10    0.800000     0.0  1.000000          0         0
11    0.600000     0.0  0.666667          1         0
12    0.600000     0.0  1.000000          0         0
13    0.600000     0.0  0.666667          0         0
14    0.628571     0.0  0.333333          0         0
15    0.657143     0.0  1.000000          1         0
16    0.542857     0.0  1.000000          0         0
17    0.600000     0.0  1.000000          0         0
18    0.600000     0.0  0.333333          0         0
19    0.685714     0.0  0.666667          0         0
20    0.771429     0.0  1.000000          0         0
21    0.742857     0.0  1.000000          0         0
22    0.857143     0.0  0.000000          1         0
23    0.685714     0.0  1.000000          1         0
24    0.742857     0.0  1.000000          0         0
25    0.885714     0.0  0.666667          0         0
26    0.742857     0.0  1.000000          0         0
27    0.685714     0.0  1.000000          0         0
28    0.714286     0.0  1.000000          1         0
29    0.771429     0.0  1.000000          0         0
...        ...     ...       ...        ...       ...
4670  0.857143     1.0  0.333333          0         0
4671  0.885714     1.0  0.666667          1         0
4672  0.742857     1.0  0.333333          1         0
4673  0.685714     1.0  0.333333          0         0
4674  0.714286     1.0  0.333333          0         0
4675  0.685714     1.0  1.000000          1         0
4676  0.657143     1.0  0.333333          0         0
4677  0.657143     1.0  0.666667          1         0
4678  0.714286     1.0  1.000000          0         1
4679  0.714286     1.0  0.333333          0         1
4680  0.800000     1.0  1.000000          0         1
4681  0.714286     1.0  0.333333          1         1
4682  0.828571     1.0  0.666667          0         1
4683  0.571429     1.0  1.000000          0         1
4684  0.628571     1.0  1.000000          0         1
4685  0.742857     1.0  0.333333          0         1
4686  0.742857     1.0  1.000000          0         1
4687  0.600000     1.0  1.000000          0         1
4688  0.542857     1.0  1.000000          1         1
4689  0.542857     1.0  1.000000          1         1
4690  0.571429     1.0  0.666667          1         1
4691  0.571429     1.0  0.666667          1         1
4692  0.628571     1.0  0.333333          1         1
4693  0.628571     1.0  0.333333          0         1
4694  0.514286     1.0  0.333333          0         1
4695  0.828571     1.0  0.666667          1         0
4696  0.714286     1.0  0.666667          0         1
4697  0.714286     1.0  0.333333          0         0
4698  0.800000     1.0  1.000000          0         0
4699  0.771429     1.0  0.333333          1         1

[4700 rows x 5 columns]

data transfer fee dari 0 hingga 1000 sudah mengalami proses filling null value sehingga kurang baik jika dilibatkan dalam pembuatan model

In [130]:
data = dataset2[1000:]
data

Age    Season      Role  Migration  Category
1000  0.657143  0.222222  0.666667          0         0
1001  0.514286  0.222222  1.000000          1         1
1002  0.685714  0.222222  0.666667          0         0
1003  0.628571  0.222222  0.000000          0         0
1004  0.771429  0.222222  0.333333          0         1
1005  0.685714  0.222222  1.000000          0         0
1006  0.742857  0.222222  0.333333          1         0
1007  0.628571  0.222222  0.333333          0         0
1008  0.657143  0.222222  0.666667          1         0
1009  0.714286  0.222222  0.333333          0         1
1010  0.771429  0.222222  1.000000          0         0
1011  0.714286  0.222222  1.000000          0         0
1012  0.885714  0.222222  0.333333          1         0
1013  0.571429  0.222222  0.666667          0         1
1014  0.742857  0.222222  0.333333          0         0
1015  0.857143  0.222222  0.333333          1         0
1016  0.742857  0.222222  0.666667          1         0
1017  0.714286  0.222222  1.000000          1         0
1018  0.657143  0.222222  1.000000          0         0
1019  0.800000  0.222222  0.333333          1         0
1020  0.742857  0.222222  1.000000          1         0
1021  0.800000  0.222222  1.000000          0         0
1022  0.771429  0.222222  0.666667          1         0
1023  0.742857  0.222222  1.000000          1         0
1024  0.742857  0.222222  0.666667          0         0
1025  0.657143  0.222222  0.666667          0         0
1026  0.742857  0.222222  1.000000          0         0
1027  0.571429  0.222222  1.000000          0         0
1028  0.685714  0.222222  0.333333          1         0
1029  0.771429  0.222222  1.000000          0         0
...        ...       ...       ...        ...       ...
4670  0.857143  1.000000  0.333333          0         0
4671  0.885714  1.000000  0.666667          1         0
4672  0.742857  1.000000  0.333333          1         0
4673  0.685714  1.000000  0.333333          0         0
4674  0.714286  1.000000  0.333333          0         0
4675  0.685714  1.000000  1.000000          1         0
4676  0.657143  1.000000  0.333333          0         0
4677  0.657143  1.000000  0.666667          1         0
4678  0.714286  1.000000  1.000000          0         1
4679  0.714286  1.000000  0.333333          0         1
4680  0.800000  1.000000  1.000000          0         1
4681  0.714286  1.000000  0.333333          1         1
4682  0.828571  1.000000  0.666667          0         1
4683  0.571429  1.000000  1.000000          0         1
4684  0.628571  1.000000  1.000000          0         1
4685  0.742857  1.000000  0.333333          0         1
4686  0.742857  1.000000  1.000000          0         1
4687  0.600000  1.000000  1.000000          0         1
4688  0.542857  1.000000  1.000000          1         1
4689  0.542857  1.000000  1.000000          1         1
4690  0.571429  1.000000  0.666667          1         1
4691  0.571429  1.000000  0.666667          1         1
4692  0.628571  1.000000  0.333333          1         1
4693  0.628571  1.000000  0.333333          0         1
4694  0.514286  1.000000  0.333333          0         1
4695  0.828571  1.000000  0.666667          1         0
4696  0.714286  1.000000  0.666667          0         1
4697  0.714286  1.000000  0.333333          0         0
4698  0.800000  1.000000  1.000000          0         0
4699  0.771429  1.000000  0.333333          1         1

[3700 rows x 5 columns]

# Splitting & Model Selection

In [131]:
array = data.values
x = array[:, 0:4]
y = array[:, 4]
num_data_points = len(y)

x_train = x[:int(.65 * num_data_points)]
y_train = y[:int(.65 * num_data_points)]
x_test = x[int(.65 * num_data_points):int(.85 * num_data_points)]
y_test = y[int(.65 * num_data_points):int(.85 * num_data_points)]
cross_val_x = x[int(.85 * num_data_points):]
cross_val_y = y[int(.85 * num_data_points):]

# Deploying Model & Result

Di sini digunakan Logistic Regression untuk melihat tingkat akurasi model dalam menebak kelas 

In [132]:
model = LogisticRegression()
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

print('Logistic Regression accuracy (Test): %f' % model.score(x_test, y_test))
print('Logistic Regression accuracy (Validation): %f' % model.score(cross_val_x, cross_val_y))

Logistic Regression accuracy (Test): 0.741892
Logistic Regression accuracy (Validation): 0.744144


C:\Users\Bangga Rakana Adian\Anaconda3\envs\mentoring_rm_3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


# Summary

1. Kategori overpriced pemain sepak bola dapat diklasifikasi berdasarkan feature posisi pemain, usia, musim, dan jenis migrasi pemain (transfer domestik atau lintas liga). 
2. Performa dari model diuji dengan dua data: data testing dan data validation dengan masing-masing sebesar 20% dan 15% dari keseluruhan data
3. Feature Engineering digunakan untuk menerjemahkan feature yang semula tidak bisa dikalkulasikan sehingga feature tersebut dapat dimasukkan ke dalam perancangan model
4. Akurasi model untuk masing-masing data testing dan data validation adalah 74.18% dan 74.41%

# Evaluations and Improvements

1. Seribu data pertama dalam dataset bersifat destruktif pada performa model. Sebab, feature 'Transfer Fee' mengalami terlalu banyak cleansing karena banyak yang mengandung null. Null value tersebut diisi sama persis dengan nilai pada feature 'Market Value'
2. Seribu data pertama tidak include dalam pembuatan model sehingga total row data yang dapat diolah sebesar 3699
3. Untuk pengembangan, dataset ini memerlukan data tambahan yang lebih spesifik untuk meningkatkan akurasi prediksi, contohnya seperti jumlah pertandingan yang dimainkan, jumlah gol, jumlah assist, dan jumlah cleansheet (jika seorang goalkeeper).
